File: Companies/Amazon/HammingDistance.java
package Companies.Amazon;

/**
 * Hamming distance between two strings of equal length is the number of positions
 * at which the corresponding symbols are different, exp: 1011101 and 1001001 is 2.
 */
public class HammingDistance {
    public int hammingDistance(int x, int y) {
//        return Integer.bitCount(x^y); // count number of bit '1'

        int xor = x^y;
        int dis = 0;
        while (xor != 0) {
            if (xor%2 == 1) {
                dis++;
            }
            xor = xor >> 1;
        }
        return dis;
    }

    /**
     * Remove rightmost bit of '1'
     * x = 10001000, x-1 = 10000111, x&(x-1)=10000000
     */
    public int hammingDistanceII(int x, int y) {
        int xor = x^y;
        int dis = 0;
        while (xor != 0) {
            dis++;
            xor = xor&(xor-1);
        }
        return dis;
    }
}

File: Topics/Design/al/HammingDistance.cpp
class Solution {
public:
    /**
     * Calculate the Hamming distance between two integers.
     * The Hamming distance is the number of positions at which 
     * the corresponding bits are different.
     * 
     * @param x First integer
     * @param y Second integer
     * @return The number of bit positions where x and y differ
     */
    int hammingDistance(int x, int y) {
        // XOR operation highlights the differing bits (sets them to 1)
        // __builtin_popcount counts the number of set bits (1s)
        return __builtin_popcount(x ^ y);
        // return __builtin_popcount(~(x ^ y)); 
    }

    int hammingDistance(int x, int y) {
        int diff = x ^ y;
        int count = 0;
    
        while (diff) {
            diff &= (diff - 1);
            ++count;
        }
    
        return count;
    }
    
};
